File: src/furaffinity/pages/submission.rs
use anyhow::{Context, Result};
use scraper::Html;

use super::html_wrapper::HtmlElement;
const SUBMISSION_NOT_FOUND_TEXT: &str =
    "The submission you are trying to find is not in our database.";
const UNAUTHENTICATED_TEXT: &str = "please log in or create an account";
const CLOUDFLARE_JS_CHALLENGE_SELECTOR: &str = "#challenge-form";
const CLOUDFLARE_JS_REQUIRED_TEXT: &str = "Enable JavaScript and cookies to continue";
const CLOUDFLARE_CHECKING_TEXT: &str = "checking your browser";

#[derive(Debug)]
pub enum SubmissionPageVariant {
    NotFound,
    Unauthenticated,
    Blocked,
    ImageSubmission,
    FlashSubmission,
}

pub struct SubmissionPage<'a>(HtmlElement<'a>);
impl<'a> SubmissionPage<'a> {
    pub fn new(html: &'a Html) -> Self {
        SubmissionPage(HtmlElement::from(html.root_element()))
    }

    pub fn get_variant(&self) -> SubmissionPageVariant {
        if self.0.select(CLOUDFLARE_JS_CHALLENGE_SELECTOR).is_ok() {
            return SubmissionPageVariant::Blocked;
        }

        let page_text = self.0.text().to_lowercase();
        if page_text.contains(&CLOUDFLARE_JS_REQUIRED_TEXT.to_lowercase())
            || page_text.contains(CLOUDFLARE_CHECKING_TEXT)
        {
            return SubmissionPageVariant::Blocked;
        }

        let section_body = self.0.select(".section-body");
        if let Ok(section_body) = section_body {
            if section_body.text().contains(SUBMISSION_NOT_FOUND_TEXT) {
                return SubmissionPageVariant::NotFound;
            }

            if section_body.text().contains(UNAUTHENTICATED_TEXT) {
                return SubmissionPageVariant::Unauthenticated;
            }
        }

        if self.0.select("#flash_embed").is_ok() {
            return SubmissionPageVariant::FlashSubmission;
        }

        SubmissionPageVariant::ImageSubmission
    }

    pub fn get_url(&self) -> Result<String> {
        self.0.select("meta[property*='og:url']")?.attr("content")
    }

    pub fn get_title(&self) -> Result<String> {
        self.0.select("meta[property*='og:title']")?.attr("content")
    }

    pub fn get_description(&self) -> Result<String> {
        self.0
            .select("meta[property*='og:description']")?
            .attr("content")
    }

    pub fn get_view_count(&self) -> Result<usize> {
        self.0
            .select("div.views span")?
            .text()
            .parse()
            .with_context(|| "Could not parse view count")
    }

    pub fn get_comment_count(&self) -> Result<usize> {
        self.0
            .select("section.stats-container div.comments span")?
            .text()
            .parse()
            .with_context(|| "Could not parse comment count")
    }

    pub fn get_fave_count(&self) -> Result<usize> {
        self.0
            .select("div.favorites span")?
            .text()
            .parse()
            .with_context(|| "Could not parse fave count")
    }

    pub fn get_submission_download_url(&self) -> Result<String> {
        Ok(format!(
            "https:{}",
            self.0.select("div.download a")?.attr("href")?
        ))
    }

    pub fn get_thumbnail_download_url(&self) -> Result<String> {
        Ok(format!(
            "https:{}",
            self.0.select("#submissionImg")?.attr("data-preview-src")?
        ))
    }
}

File: src/furaffinity/submission.ts
import * as cheerio from 'cheerio';
import { stripBbcode } from './bbcode.js';
import { cleanDescriptionHtml, plainToHtml } from './description.js';
import type { SubmissionInfo } from './submissionInfo.js';

type SubmissionPageInfo = Omit<SubmissionInfo, 'sizeBytes' | 'contentType' | 'width' | 'height'>;
type CommonPageInfo = Omit<SubmissionPageInfo, 'imageUrl'>;
type StoryPageInfo = CommonPageInfo & { contentUrl: string; extension: string };
type MusicPageInfo = CommonPageInfo & { audioUrl: string };

export type SubmissionPageResult =
  | { type: 'image'; info: SubmissionPageInfo }
  | { type: 'story'; info: StoryPageInfo }
  | { type: 'music'; info: MusicPageInfo }
  | { type: 'flash' }
  | { type: 'notFound' }
  | { type: 'unauthenticated' }
  | { type: 'accountDisabled' }
  | { type: 'blocked' }
  | { type: 'temporarilyOffline' };

const SUBMISSION_NOT_FOUND_TEXT = 'not in our database';
const UNAUTHENTICATED_TEXT = 'please log in';
const LOGIN_REQUIRED_TITLE = 'login required';
const ACCOUNT_DISABLED_TITLE = 'account disabled';
const FA_OFFLINE_TITLE = 'temporarily offline';
const CLOUDFLARE_JS_REQUIRED_TEXT = 'enable javascript and cookies to continue';
const CLOUDFLARE_CHECKING_TEXT = 'checking your browser';

export function parseSubmissionPage(html: string): SubmissionPageResult {
  const $ = cheerio.load(html);

  if ($('#challenge-form').length > 0) {
    return { type: 'blocked' };
  }

  const pageText = $.root().text().toLowerCase();
  if (pageText.includes(CLOUDFLARE_JS_REQUIRED_TEXT) || pageText.includes(CLOUDFLARE_CHECKING_TEXT)) {
    return { type: 'blocked' };
  }

  // Check for login required page (expired/invalid session)
  const pageTitle = $('title').text().toLowerCase();
  if (pageTitle.startsWith(LOGIN_REQUIRED_TITLE)) {
    return { type: 'unauthenticated' };
  }
  if (pageTitle.startsWith(ACCOUNT_DISABLED_TITLE)) {
    return { type: 'accountDisabled' };
  }
  if (pageTitle.includes(FA_OFFLINE_TITLE)) {
    return { type: 'temporarilyOffline' };
  }

  const sectionBodyText = $('.section-body').text();
  if (sectionBodyText.includes(SUBMISSION_NOT_FOUND_TEXT)) {
    return { type: 'notFound' };
  }
  if (sectionBodyText.toLowerCase().includes(UNAUTHENTICATED_TEXT)) {
    return { type: 'unauthenticated' };
  }

  if ($('#flash_embed').length > 0) {
    return { type: 'flash' };
  }

  const url = $('meta[property*="og:url"]').attr('content');
  const title = $('meta[property*="og:title"]').attr('content');
  const rawDescription = $('meta[property*="og:description"]').attr('content');
  const viewCountText =
    $('.submission-page-stats div[title="Views"] div').first().text().trim() || $('div.views span').first().text(); // fallback old layout
  const commentCountText =
    $('.submission-page-stats div[title="Comments"] div').first().text().trim() ||
    $('section.stats-container div.comments span').first().text();
  const faveCountText =
    $('.submission-page-stats div[title="Favorites"] div').first().text().trim() ||
    $('div.favorites span').first().text();
  const thumbnailSrc = $('#submissionImg').attr('data-preview-src');
  const newArtistLink = $('.submission-description-artist .c-usernameBlockSimple a[href^="/user/"]').first();
  const artistLink = newArtistLink.length ? newArtistLink : $('.submission-id-sub-container a[href^="/user/"]').first(); // fallback old layout
  const artistName = artistLink.text().trim();
  const artistHref = artistLink.attr('href');

  if (!url || !title || rawDescription === undefined || !thumbnailSrc || !artistName || !artistHref) {
    throw new Error(
      `Failed to parse submission page: missing fields (url=${url}, title=${title}, description=${rawDescription}, thumbnail=${thumbnailSrc}, artist=${artistName})`,
    );
  }

  const description = stripBbcode(rawDescription);
  const descriptionTextEl = $('.submission-description-text').first();
  const descriptionHtml = descriptionTextEl.length
    ? cleanDescriptionHtml(descriptionTextEl.html() ?? '')
    : plainToHtml(description);

  const viewCount = parseInt(viewCountText, 10);
  const commentCount = parseInt(commentCountText, 10);
  const faveCount = parseInt(faveCountText, 10);

  if (Number.isNaN(viewCount) || Number.isNaN(commentCount) || Number.isNaN(faveCount)) {
    throw new Error(
      `Failed to parse submission counts (views="${viewCountText}", comments="${commentCountText}", faves="${faveCountText}")`,
    );
  }

  const commonInfo: CommonPageInfo = {
    url,
    title,
    description,
    descriptionHtml,
    viewCount,
    commentCount,
    faveCount,
    thumbnailUrl: `https:${thumbnailSrc}`,
    artistName,
    artistUrl: `https://www.furaffinity.net${artistHref}`,
  };

  const script = $('script[data-content-type]');
  const faContentType = script.attr('data-content-type');

  if (faContentType === 'music') {
    const audioSrc = $('#c-musicPlayer_inner').attr('src');
    if (!audioSrc) {
      throw new Error('Failed to parse music submission: missing audio URL');
    }
    return { type: 'music', info: { ...commonInfo, audioUrl: `https:${audioSrc}` } };
  }

  if (faContentType === 'text') {
    const contentUrl = script.attr('data-content-url');
    const extension = script.attr('data-content-extension');
    if (!contentUrl || !extension) {
      throw new Error('Failed to parse story submission: missing content URL or extension');
    }
    return { type: 'story', info: { ...commonInfo, contentUrl: `https:${contentUrl}`, extension } };
  }

  // New FA layout: Download button in #submission-options
  const downloadHref =
    $('#submission-options a[href^="//d.furaffinity.net"]').attr('href') || $('div.download a').attr('href'); // fallback old layout
  if (!downloadHref) {
    throw new Error(`Failed to parse submission page: missing download href (url=${url})`);
  }

  return {
    type: 'image',
    info: { ...commonInfo, imageUrl: `https:${downloadHref}` },
  };
}
